### PR TITLE
Felix/regression

### PR DESCRIPTION
**Pull request by fleibfried**

### CODE REVIEW

File: bayesian_benchmarks/tasks/classification.py [hunk 1]
```diff
@@ -32,35 +33,60 @@ def run(ARGS, data=None, model=None, is_test=False):
...
+        return np.eye(K)[Y.flatten().astype(int)].reshape(Y.shape[:-1] + (K,))
 
-    Y_oh = onehot(data.Y_test, data.K)[None, :, :]  # 1, N_test, K
+    Y_oh = onehot(data.Y_test, data.K)[None, :, :]  # [1 x N_test x K]
 
     model.fit(data.X_train, data.Y_train)
-    p = model.predict(data.X_test)  # N_test, K
+    p = model.predict(data.X_test)  # [N_test x K] or [samples x N_test x K]
+
+    assert len(p.shape) in {2, 3}  # 3-dim in case of approximate predictions (multiple samples per each X)
```
Comment (vdutor) [inline bayesian_benchmarks/tasks/classification.py]:
```suggestion
    assert p.ndim in {2, 3}  # 3-dim in case of approximate predictions (multiple samples per each X)
```

File: bayesian_benchmarks/tasks/classification.py [hunk 2]
```diff
@@ -32,35 +33,60 @@ def run(ARGS, data=None, model=None, is_test=False):
...
+
+        logp = multinomial.logpmf(Y_oh, n=1, p=p)
+
+        res['test_loglik'] = np.average(logp)
+
+        pred = np.argmax(p, axis=-1)
+
+        res['test_acc'] = np.average(np.array(pred == data.Y_test.flatten()).astype(float))
+
+        res['Y_test'] = data.Y_test
```
Comment (vdutor) [inline bayesian_benchmarks/tasks/classification.py]: why do we need to store `res["Y_test"]` in the results?

Comment (vdutor) [reply bayesian_benchmarks/tasks/classification.py]: you add Y_test and p to res for both p.shape == 2 and p.shape == 3. Better to take it out

File: bayesian_benchmarks/tasks/classification.py [hunk 3]
```diff
@@ -32,35 +33,60 @@ def run(ARGS, data=None, model=None, is_test=False):
...
     p = np.clip(p, eps, 1 - eps)
     p = p / np.expand_dims(np.sum(p, -1), -1)
 
+    assert np.all(p >= 0.0) and np.all(p <= 1.0)
+
     # evaluation metrics
     res = {}
 
-    logp = multinomial.logpmf(Y_oh, n=1, p=p)
+    if len(p.shape) == 2:  # keep analysis as in the original code in case 2-dim predictions
```
Comment (vdutor) [inline bayesian_benchmarks/tasks/classification.py]: using `p.ndim` is more concise here

File: bayesian_benchmarks/tasks/classification.py [hunk 4]
```diff
@@ -32,35 +33,60 @@ def run(ARGS, data=None, model=None, is_test=False):
...
+
+        logp = multinomial.logpmf(Y_oh, n=1, p=p)
```
Comment (vdutor) [inline bayesian_benchmarks/tasks/classification.py]: adding shapes here would be useful. I believe logp is [N] here?

File: bayesian_benchmarks/tasks/classification.py [hunk 5]
```diff
@@ -32,35 +33,60 @@ def run(ARGS, data=None, model=None, is_test=False):
...
+        res['Y_test'] = data.Y_test
 
-    res['test_loglik'] = np.average(logp)
+        for n in range(p.shape[0]):  # iterate through samples
+            logp = multinomial.logpmf(Y_oh, n=1, p=p[n])
+            res['test_loglik'].append(logp)
 
-    pred = np.argmax(p, axis=-1)
+        # Mixture test likelihood (mean over per data point evaluations)
+        logp = logsumexp(res['test_loglik'], axis=0) - np.log(p.shape[0])
```
Comment (vdutor) [inline bayesian_benchmarks/tasks/classification.py]: Personally I would create a helper function `meansumexp`, you can use it in the other file as well

**Review by vdutor — CHANGES_REQUESTED** [review]

File: bayesian_benchmarks/tasks/classification.py (hunk 2 above)
Comment (fleibfried) [reply bayesian_benchmarks/tasks/classification.py]: I just followed Hugh's implementation. I don't know why he added it.

File: bayesian_benchmarks/tasks/classification.py [hunk 6]
```diff
@@ -32,35 +33,55 @@ def run(ARGS, data=None, model=None, is_test=False):
...
+        # Mixture test likelihood (mean over per data point evaluations)
+        res['test_loglik'] = meansumexp(res['test_loglik'])
 
-    pred = np.argmax(p, axis=-1)
+        p = np.mean(p, axis=0)
+        pred = np.argmax(p, axis=-1)
 
     res['test_acc'] = np.average(np.array(pred == data.Y_test.flatten()).astype(float))
 
     res['Y_test'] = data.Y_test
```
Comment (vdutor) [inline bayesian_benchmarks/tasks/classification.py]: I think it is worth figuring out, and deleting them if we don't use them anywhere. Storing 2 arrays of length N in our database is not a good idea anyway.

File: bayesian_benchmarks/tasks/utils.py [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+Helper functions...
+
+"""
+
+from typing import List
+
+import numpy as np
+from scipy.special import logsumexp
+
+def meansumexp(logps: List[np.ndarray]) -> np.ndarray:
```
Comment (vdutor) [inline bayesian_benchmarks/tasks/utils.py]: Given that this is a helper function I would make it more general by letting it accept an array (instead of a List of array's) and an axis on which to reduce. Namely,
`meansumexp(a: np.array, axis)`

File: bayesian_benchmarks/tasks/regression.py [hunk 8]
```diff
@@ -28,26 +29,66 @@ def run(ARGS, data=None, model=None, is_test=False):
...
+
+    assert len(m.shape) == len(v.shape)
+    assert len(m.shape) in {2, 3}  # 3-dim in case of approximate predictions (multiple samples per each X)
+    assert np.all(v >= 0.0)
 
     res = {}
+    log_eps = np.log(1e-12)  # log probability threshold
+    log_1_minus_eps = np.log(1.0 - 1e-12)
+
+    if len(m.shape) == 2:  # keep analysis as in the original code in case of 2-dim predictions
```
Comment (vdutor) [inline bayesian_benchmarks/tasks/regression.py]: same with `ndim`

File: bayesian_benchmarks/tasks/regression.py [hunk 9]
```diff
@@ -28,26 +29,66 @@ def run(ARGS, data=None, model=None, is_test=False):
     model = model or get_regression_model(ARGS.model)(is_test=is_test, seed=ARGS.seed)
 
     model.fit(data.X_train, data.Y_train)
-    m, v = model.predict(data.X_test)
+    m, v = model.predict(data.X_test)  # both [data points x output dim] or [samples x data points x output dim]
+
+    assert len(m.shape) == len(v.shape)
+    assert len(m.shape) in {2, 3}  # 3-dim in case of approximate predictions (multiple samples per each X)
```
Comment (vdutor) [inline bayesian_benchmarks/tasks/regression.py]: `ndim` is cleaner here

File: tests/test_tasks.py [hunk 10]
```diff
@@ -77,3 +78,118 @@ def test_active_learning_discrete(model, dataset):
...
+# Below two tests, one for regression and one for classification (2-dim and 3-dim case respectively)
+
+regression_tuple = (RegressionDataMock(), RegressionMock(), regression_results)
+approx_regression_tuple = (RegressionDataMock(), ApproximateRegressionMock(), approximate_regression_results)
+
+@pytest.mark.parametrize('tuple', [regression_tuple, approx_regression_tuple])
+def test_regression(tuple):
+    data, model, correct_result = tuple
+    result = run_regression(None, data=data, model=model, is_test=True)
+    assert correct_result['test_loglik'] == pytest.approx(result['test_loglik'], 1e-3)
```
Comment (vdutor) [inline tests/test_tasks.py]: I think most of us use `np.testing.assert_almost_equal` instead of `pytest.approx`. `np.testing` gives nice and clear errors when the assert isn't fulfilled.

Comment (vdutor) [inline tests/test_tasks.py]: Instead of the copying here I would write this as a for loop over 
```
for (ref_key, ref_value), (actual_key, actual_value) in zip(correct_result.items(), result.items()):
     np.testing.assert_equal(ref_key, actual_key)
     np.testing.assert_almost_equal(ref_value, actual_value)

```

File: tests/test_tasks.py [hunk 11]
```diff
@@ -77,3 +78,118 @@ def test_active_learning_discrete(model, dataset):
...
+approx_classification_tuple = (ClassificationDataMock(), ApproximateClassificationMock(), approximate_classification_results)
+
+@pytest.mark.parametrize('tuple', [classification_tuple, approx_classification_tuple])
+def test_classification(tuple):
+    data, model, correct_result = tuple
+    result = run_classification(None, data=data, model=model, is_test=True)
+    assert correct_result['test_loglik'] == pytest.approx(result['test_loglik'], 1e-3)
+    assert correct_result['test_acc'] == pytest.approx(result['test_acc'], 1e-3)
+    assert np.allclose(correct_result['Y_test'], result['Y_test'], rtol=0.0, atol=1e-3)
+    assert np.allclose(correct_result['p_test'], result['p_test'], rtol=0.0, atol=1e-3)
```
Comment (vdutor) [inline tests/test_tasks.py]: nitpick: you want to newline at the end of the file.

File: tests/test_tasks.py [hunk 12]
```diff
@@ -77,3 +78,118 @@ def test_active_learning_discrete(model, dataset):
...
+        mu = np.array([[1., 2., 3.], [4., 5., 6.]])
+        var = np.array([[.1, .2, .3], [.4, .5, .6]])
+        return mu, var
+
+class ApproximateRegressionMock(RegressionMock):
+    """
+    Approximate regression mock.
+    """
+    def predict(self, X: np.ndarray) -> (np.ndarray, np.ndarray):
+        mu = np.array([[[1., 2., 3.], [4., 5., 6.]], [[1.5, 2.5, 3.5], [4.5, 5.5, 6.5]]])
```
Comment (vdutor) [inline tests/test_tasks.py]: can you explain in the code why these classes return these specific shapes, please.

**Review by vdutor — CHANGES_REQUESTED** [review]

File: bayesian_benchmarks/tasks/utils.py [hunk 13]
```diff
@@ -0,0 +1,18 @@
...
+from scipy.special import logsumexp
+
+def meanlogsumexp(logps: Union[List[np.ndarray], np.ndarray], axis: int=0) -> np.ndarray:
+    """
+    Mean log sum exp of a log p list.
+    :param logps: list of log probs [samples x data points] or [samples x data points x output dim]
+    :param axis: determines reduction
+    :return: avg probability value [1]
+    """
+    return np.mean(logsumexp(logps, axis=axis) - np.log(len(logps)))
```
Comment (vdutor) [inline bayesian_benchmarks/tasks/utils.py]: `np.log(len(logps))` doesnt work `np.log(logp.shape[axis])`

**Review by vdutor — APPROVED** [review]
